schnorr_fun/src/frost/chilldkg.rs

```diff
@@ -0,0 +1,1256 @@
+        /// The input from the inner protocol
+        pub inner: simplepedpop::KeygenInput,
+        /// The shares encrypted for each receiving party
+        pub encrypted_shares: BTreeMap<PartyIndex, Scalar<Public, Zero>>,
+        /// The multi-encryption nonce for the encryptions in `encrypted_shares`
+        pub encryption_nonce: Point,
+    }
+
+    /// Stores the state of the coordinator as it aggregates inputs from [`Contributor`]s.
+    #[derive(Clone, Debug, PartialEq)]
+    pub struct Coordinator {
+        inner: simplepedpop::Coordinator,
+        agg_encrypted_shares: BTreeMap<PartyIndex, (Point, Scalar<Public, Zero>)>,
+        encryption_nonces: Vec<Point>,
+    }
+
+    impl Coordinator {
+        /// Creates a new coordinator with:
+        ///
+        /// - `threshold`: of key we're trying to generate
+        /// - `n_contribtors`: The number of [`Contributor`]s
+        /// - `receiver_encryption_keys`: The encryption keys of each of the share receivers.
+        pub fn new(
+            threshold: u32,
+            n_contribtors: u32,
+            receiver_encryption_keys: &BTreeMap<PartyIndex, Point>,
+        ) -> Self {
+            let agg_encrypted_shares = receiver_encryption_keys
+                .iter()
+                .map(|(&receiver, encryption_key)| (receiver, (*encryption_key, Scalar::zero())))
+                .collect();
+            Self {
+                inner: simplepedpop::Coordinator::new(threshold, n_contribtors),
+                agg_encrypted_shares,
+                encryption_nonces: Default::default(),
+            }
+        }
+
+        /// Adds an `input` from a [`Contributor`].
+        ///
+        /// Note verifying this is the correct input from the correct party is up to your application!
+        pub fn add_input<H: Hash32, NG>(
+            &mut self,
+            schnorr: &Schnorr<H, NG>,
+            from: u32,
+            input: KeygenInput,
+        ) -> Result<(), &'static str> {
+            if self.inner.is_finished() {
+                return Err("all inputs have already been collected");
+            }
+            let mut check_missing = self.agg_encrypted_shares.keys().collect::<BTreeSet<_>>();
+
+            for dest in input.encrypted_shares.keys() {
+                if !self.agg_encrypted_shares.contains_key(dest) {
+                    return Err("included share for unknown party");
+                }
+                check_missing.remove(dest);
+            }
+
+            if !check_missing.is_empty() {
+                return Err("didn't have share for all parties");
+            }
+
+            // ⚠ only do mutations after we're sure everything is ok
+            self.inner.add_input(schnorr, from, input.inner)?;
+
+            for (dest, encrypted_share_contrib) in input.encrypted_shares {
+                let agg_encrypted_share = &mut self.agg_encrypted_shares.get_mut(&dest).unwrap().1;
+                *agg_encrypted_share += encrypted_share_contrib;
```

Comment:
I was going to suggest doing this aggregation in `finish` but actually it's good here since you can reuse simplepedpop types nicely

Reply:
The simpler state is better.